library/std/src/sys/sync/once/queue.rs

```diff
@@ -140,6 +140,16 @@ impl Once {
         }
     }
 
+    #[inline]
+    pub(crate) fn set_state(&mut self, new_state: ExclusiveState) {
+        *self.state_and_queue.get_mut() =
+            self.state_and_queue.get_mut().with_addr(match new_state {
+                ExclusiveState::Incomplete => INCOMPLETE,
+                ExclusiveState::Poisoned => POISONED,
+                ExclusiveState::Complete => COMPLETE,
+            });
```

Comment:
The `with_addr` confused me, there is no reason to carry over the provenance.
```suggestion
        *self.state_and_queue.get_mut() = match new_state {
            ExclusiveState::Incomplete => ptr::without_provenance_mut(INCOMPLETE),
            ExclusiveState::Poisoned => ptr::without_provenance_mut(POISONED),
            ExclusiveState::Complete => ptr::without_provenance_mut(COMPLETE),
        };
```